### DISCUSSION TITLE
Welcome to WalletConnectUnity Discussions!

### DISCUSSION THREAD

**Discussion by ecp4224** [WalletConnect/WalletConnectUnity#22] [Announcements]
## 👋 Welcome to WalletConnectUnity!
  Since we are all Unity developers and we all using this library to either create awesome games or experiences, I thought it'd be a good idea to have a shared place to ask questions, share content and discuss things about WalletConnect in games and WalletConnectUnity in general. We’re using Discussions as a place to connect with other members of the WalletConnectUnity community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas or game projects you are working on.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you are using this library for.

**Comment by JoshOtter** [comment]
Hello, my name is Josh, and I am excited about this library for use with Unity. I found it while looking for a way to integrate WalletConnect in a project using the Algorand blockchain. I've been able to create transactions in Unity using this [Dotnet-Algorand-SDK](https://github.com/RileyGe/dotnet-algorand-sdk), but I want to learn how to utilize WalletConnectUnity to further secure the transactions without requiring the user to enter their private key. There is a lot of code to look through and work with, so I'm sure I'll have questions along the way. Thank you for your work on this project!

**Reply by ecp4224** [reply]
From my limited knowledge of Algorand you should be able to utilize this. I'm not sure if this will help, but this is the [WalletConnectSharp NEthereum extension](https://github.com/WalletConnect/WalletConnectSharp/blob/main/WalletConnectSharp.NEthereum/Client/WalletConnectClient.cs). 

This extension acts as a web3 provider by taking the raw JSON RPC requests that NEthereum's web3 creates and pushes them through WalletConnect and awaits a reply to give back to NEthereum's web3. From reading [Algorand's docs about using WalletConnect in js](https://developer.algorand.org/docs/get-details/walletconnect/), you will most likely want to send custom JSON RPC requests as well. In their docs, the line

```
const result: Array<string | null> = await this.connector.sendCustomRequest(request);
```

Is the same as

```
MyCustomRpcRequestMessage rpcRequestMessage = GenerateAlgorandJsonRpcMessage();

MyCustomRpcResponseMessage response = WalletConnect.ActiveSession.SendRequestAwaitResponse(rpcRequestMessage, rpcRequestMessage.Id);
``` 

Let me know if you have any question I'll be more than happy to help

**Reply by JoshOtter** [reply]
Thank you for your feedback! I am currently working through the WalletConnectUnity code to see how everything works. I'm sure I'll have some questions for you as I go, so I appreciate your willingness to help!

**Reply by JoshOtter** [reply]
Hi, I have been looking through all the WalletConnectUnity code and I'm getting ready to try to test some transactions. But first, I was hoping to see if it is possible to access data about assets amounts from a wallet through WalletConnectUnity. For example, I would like a user to be able to see in the game how much of a particular asset they have available to make purchases. Maybe I missed it, but I haven't found anything, even on WalletConnect's documentation, that would seem to support this capability. Do you know if this is possible?

**Reply by JoshOtter** [reply]
Hello again. I've figured out my previous question. I got the wallet's public key from WalletConnect.ActiveSession.Accounts[0], and I used it with the Dotnet Algorand SDK I mentioned in my first comment to get access to its asset information.

Now I'm trying to figure out how to get an asset transfer transaction to work, and I'm having a lot of trouble figuring out how to set up the parameters to work with a MyAlgo wallet. I can find examples of them in other WalletConnect SDKs, but I'm having trouble figuring out how to transfer that information into the context of C#. I tried making some Algorand versions of the Ethereum classes, like an AlgoSendTransaction, AlgoResponse, and a TransactionData class based on Algorand, where I changed its properties to try to match what I was seeing in the examples from other SDKs. But so far that happens is I get an error that says "Invalid Input: Unable to parse transaction."

This all still very new to me, so I may be missing something simple. Unfortunately, the sample code you posted went over my head. If you have any additional advice or insight, I would greatly appreciate it.

**Reply by JoshOtter** [reply]
This morning I tried sending an [Algorand Transaction](https://github.com/RileyGe/dotnet-algorand-sdk/blob/345ac4a540abb3f8094b2f97be81bcef3ff6f385/dotnet-algorand-sdk/Util/Utils.cs#L371), using the SDK I mentioned. It gets sent through my custom AlgoSignTransaction() functions in the WalletConnectSession script that take an Algorand.Transaction as a parameter. It then gets transformed into an AlgoSendTransaction, which is the same as an EthSendTransaction class where it transfers the Transaction information into the new object's parameters. Then it goes through the WalletConnectSession script's Send() function and makes it all the way to the SendRequest() function where it is serialized, encrypted, and then serialized again in a NetworkMessage object's Payload. Once that completes, it continues the Send() function until it gets to the last line "return await eventCompleted.Task;" where I get an error message that says, "Invalid Input: Unable to parse transaction."

I apologize for all the messages, but I figured more information would be helpful. Thanks!

**Reply by ecp4224** [reply]
Hi,
Sorry for the delay on this. The first question I have would be what wallet are you using to test this with? If you make it all the way down to `SendRequest` and the exception is coming from `return await eventCompleted.Task;` then that means the wallet is the one that is returning the error. The `WalletConnectUnity` code [will throw an `IOException` for any error the wallet returns as a response to a `Send` request](https://github.com/WalletConnect/WalletConnectUnity/blob/main/Assets/WalletConnectUnity/Scripts/WalletConnectSharp/WalletConnectSharp.Core/WalletConnectSession.cs#L420).

 > I tried making some Algorand versions of the Ethereum classes, like an AlgoSendTransaction, AlgoResponse, and a TransactionData class based on Algorand, where I changed its properties to try to match what I was seeing in the examples from other SDKs
 
 This is the correct path to go down to implement this correctly, I think now you need to ensure the final JSON representation of these classes match what the wallet is expecting. This usually manifests in adding `[JsonProperty("jsonvarname")] ` and `[JsonIgnore]` attributes to your properties. The `Send` and `SendRequest` functions in `WalletConnectSession` performs all the json serializing for you using the standard Newtonsoft.Json library.

**Reply by JoshOtter** [reply]
Hi,
No problem! I was using the MyAlgo, official Algorand wallet. I did what you suggested and built some custom classes to catch and send the appropriate transaction properties in a Json format. After tweaking things and doing some trial and error, it is now successfully retrieving the signed transactions through WalletConnect, and submitting them through the Dotnet Algorand SDK. Thank you for your help!

**Reply by ecp4224** [reply]
Hi,
That's great to hear! I do plan on having extensions on-top of this library to interact with the different blockchain networks (since as you noted, WalletConnect doesn't offer any specific network query abilities). Currently the only extension planned is NEthereum, however it seems Algorand could be another option as well now thanks to your developments :) 

If your project is public somewhere, I'll be happy to link to it in this repo for others to find

**Reply by JoshOtter** [reply]
Hello again! Unfortunately, the project I'm working on isn't public right now, otherwise I would be happy to share the developments we've made for the Algorand side of things. Our game is currently working with Wallet Connect via pc and mac, but we're having trouble initiating a connection to our wallet through Wallet Connect. When we build it for mobile, it jumps straight to the Wallet Connect page of the wallet app on an android phone, but nothing happens after that.

I was wondering if you have any insight into what may be going wrong. I see that in the WalletConnect script, the OpenMobileWallet() and the OpenDeepLink() methods use Application.OpenURL(ConnectURL) to initiate the connection to the wallet. Does the ConnectURL variable need to be configured differently to be used with deeplink? Also, when it does connect, does the rest of the code that handles sending transactions work with that open connection, or is there any additional considerations I need to keep in mind for how requests are being sent over the connection.

I appreciate you help!

**Reply by JoshOtter** [reply]
Actually, I got it sorted out. We were having an issue where the handshakeID wasn't being generated for some reason, but its working now and the game is able to connect via deep link.

**Comment by ghost** [comment]
Hi there!

I'm Al, and I may (or potentially not, depending on flakiness - no offence meant) be using WalletConnectUnity in an upcoming game.

The concept of WCU is great, but the implementation is still obviously very much work in progress. What would help _me_ to help _you_ would be if the library code was really well commented. Unfortunately, it's not. This makes it incredibly difficult to troubleshoot when you're up against random behaviour - and also kinda futile because your comments get lost on update.

I would also suggest adding a local websocket server to communicate with wallets safely on the same device - pushing that communication online,  encrypted or not, is just such a massive un-safe inefficiency - it should really just all be kept on-device.

Looking forward to working on this with ya!

**Reply by ecp4224** [reply]
Thank you for your support + feedback! You are correct that the lack of documentation needs to be fixed, so I'll take time today and tomorrow to go through the code and add documentation as much as I can. This will also be useful because then some documentation pages can be auto generated + annotated for better developer accessibility.

In terms of bugs and stability improvements, I'm torn between polishing this version of the library or starting heavy development on the v2 protocol beta. If you've observed strange behavior or bugs please feel free to log them in the issue tracker for my sake and I'll try to fix them.

In terms of development environment, I do agree that using the production bridges is not ideal at all, perhaps some documentation on how to setup a development bridge server is needed. However last time I looked, I couldn't find a lot of information in general on how to do this for 1.x WalletConnect, I'll look again though

Excited to see what kind of game you'll make with this 😄

**Comment by GarrettGibbs** [comment]
Hello! 
I am working as part of a team on a game that has successfully used all your great work to allow users to connect with their wallet via the QR code. Now I'm working on putting together a mobile build for the game and am trying to setup a deep link instead. I see that there are methods for OpenMobileWallet and OpenDeepLink but have not been able to get them to trigger properly. Do you have any insight on the best way to transition over to a mobile build or documentation to that effect?

Thanks so much! - Garrett
